BUG: Fix for latest matplotlib

larsoner opened the pull request:
On latest matplotlib I get (this is from MNE but we vendor your fix):
```
  File "/home/larsoner/python/mne-python/mne/fixes.py", line 1104, in _crop_colorbar
    xy = cbar._outline(new_X, new_Y)
AttributeError: 'ColorbarBase' object has no attribute '_outline'
```
See inline comments for description.

Not sure if there is a better fix or not...

larsoner commented:
Okay this is clearly not the correct fix:
```
  File "/home/larsoner/python/mne-python/mne/fixes.py", line 1118, in _crop_colorbar
    outline = cbar.outline.get_xy()
AttributeError: '_ColorbarSpine' object has no attribute 'get_xy'
```
I'm wondering if this fix is perhaps not necessary on newer matplotlib. I'll wrap this in a `hasattr` check for now. Any good examples I could run to see if it's working correctly or not?

larsoner commented:
... nope, just needed to remove the `hasattr` from the conditional I had errantly left in there.

Still interested to know if there is a good example to run to check this behavior -- maybe no adjustments are necessary anymore (or maybe this fix actually does not work!).

bthirion commented on nilearn/plotting/img_plotting.py:
```diff
@@ -228,12 +228,21 @@ def _crop_colorbar(cbar, cbar_vmin, cbar_vmax):
+        y = Y.T.reshape(-1)[ii]
+        xy = (np.column_stack([y, x])
+              if cbar.orientation == 'horizontal' else
+              np.column_stack([x, y]))
```
Thx a lot ! 

All this looks good to me, but would you mind giving a few explanations ? I'm afraid of the maintenance of this code, give that some of the computations are somewhat cryptic ;-)

larsoner replied:
The code is taken from the linked URL / diff above, i.e., it keeps whatever the old computations were

larsoner replied:
(I do not know what it does actually)

bthirion commented:
LGTM. Any other comment/review ?

KamalakerDadi commented:
With matplotlib==3.3.2, I could not reproduce the issue. Is this issue anticipated with the upcoming release?

larsoner commented:
Yes I'm on 3.4-dev

larsoner commented:
> I wonder if we could add tests. A smoke test for plotting.plot_img with colorbar=True.

It seems as though this is unnecessary -- on `master` I get:
```
FAILED nilearn/plotting/tests/test_img_plotting.py::test_plot_glass_brain - AttributeError: 'ColorbarBase' object has no attribute '_outline'
FAILED nilearn/plotting/tests/test_img_plotting.py::test_plot_stat_map - AttributeError: 'ColorbarBase' object has no attribute '_outline'
FAILED nilearn/plotting/tests/test_img_plotting.py::test_plot_stat_map_threshold_for_uint8 - AttributeError: 'ColorbarBase' object has no attribute '_outline'
FAILED nilearn/plotting/tests/test_img_plotting.py::test_plot_glass_brain_threshold_for_uint8 - AttributeError: 'ColorbarBase' object has no attribute '_outline'
FAILED nilearn/plotting/tests/test_img_plotting.py::test_save_plot - FileNotFoundError: [Errno 2] No such file or directory: '/tmp/tmpdgw_nfrx.png'
FAILED nilearn/plotting/tests/test_img_plotting.py::test_plot_stat_map_colorbar_variations - AttributeError: 'ColorbarBase' object has no attribute '_outline'
FAILED nilearn/plotting/tests/test_img_plotting.py::test_singleton_ax_dim - AttributeError: 'ColorbarBase' object has no attribute '_outline'
FAILED nilearn/plotting/tests/test_img_plotting.py::test_plot_stat_map_with_nans - AttributeError: 'ColorbarBase' object has no attribute '_outline'
FAILED nilearn/plotting/tests/test_img_plotting.py::test_plotting_functions_with_cmaps - AttributeError: 'ColorbarBase' object has no attribute '_outline'
FAILED nilearn/plotting/tests/test_img_plotting.py::test_plotting_functions_with_nans_in_bg_img - AttributeError: 'ColorbarBase' object has no attribute '_outline'
FAILED nilearn/plotting/tests/test_img_plotting.py::test_plotting_functions_with_display_mode_tiled - AttributeError: 'ColorbarBase' object has no attribute '_outline'
FAILED nilearn/plotting/tests/test_img_plotting.py::test_plot_glass_brain_colorbar_having_nans - AttributeError: 'ColorbarBase' object has no attribute '_outline'
```
On this PR these pass.

I also:

1. Changed some `try/except/os.remove` patterns to use the `tmpdir` pytest fixture. It's both more standard for testing (nowadays) and nicer because it automatically cleans up.
2. Changed some `np.bool`, `np.int`, and `np.float` to their builtin equivalents because of warnings in latest `numpy` like:
    ```
    /home/larsoner/python/nilearn/nilearn/plotting/find_cuts.py:275: DeprecationWarning: `np.float` is a deprecated alias for the builtin `float`. Use `float` by itself, which is identical in behavior, to silence this warning. If you specifically wanted the numpy scalar type, use `np.float_` here.
    ```
3. Added an `autouse=True` fixture at the `plotting/tests/conftest.py` scope so that plots always close -- at one point with the failures above there were dozens of windows open on my system

larsoner commented:
... as a side note, I'd also recommend adding this to `plotting/tests/conftest.py`:
```
def pytest_configure(config):
    import matplotlib
    matplotlib.use('Agg', force=True)
```
It makes it so that no plots pop up, which is really nice (avoids focus stealing on window managers that aren't great at it).

KamalakerDadi commented:
>On this PR these pass.

Alright. Makes sense.

KamalakerDadi commented:
>Added an autouse=True fixture at the plotting/tests/conftest.py scope so that plots always close -- at one point with the failures above there were dozens of windows open on my system

That's nice. But, we already have ```plt.close()``` in each test function in order to address this issue. 

- Do you think it's not effective?
- If not, may be we remove ```plt.close()``` and use ```close_all()```? WDYT

Slightly we are deviating from the main point.  Fixing ```_outline``` is good for now. Will there be an alternative once 3.4.0 matplotlib is released?

larsoner commented:
> That's nice. But, we already have plt.close() in each test function in order to address this issue.
>
> - Do you think it's not effective?

Yes -- these only get executed if the test passes (this is point (3) I made above). The `autouse=True` is a better solution to the problem:

1. It's always executed at the end regardless of the test passing or not. It's like wrapping the entire test in a `try/except: plt.close('all')`.
2. In principle we could now delete all of the `plt.close()`s at the end of your tests to make them more DRY, which is a good thing. I can do that if you want more `-` lines in this PR.

> - If not, may be we remove plt.close() and use close_all()? WDYT

Calling `plt.close('all')` when everything is already closed doesn't add any appreciable overhead:
```
In [1]: import matplotlib.pyplot as plt                                                                                                                                             
In [2]: %timeit plt.close('all')                                                                                                                                                    
516 ns ± 2.13 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)
```
So there is no real *need* to do it in this PR. But I can do it if you want, it's easy enough.

> Will there be an alternative once 3.4.0 matplotlib is released?

I'm not sure what you're asking. Alternative to what? The code that is in this PR should work on 3.3 and 3.4. For me locally it works on 3.3, and your CIs at least are happy with it on < 3.4.

jeromedockes commented:
+1 for setting a non-gui matplotlib backend for tests, and for `plt.close("all")`  in an autoused fixture on my side.

KamalakerDadi commented:
:thumbsup:

larsoner commented:
> +1 for setting a non-gui matplotlib backend for tests, and for plt.close("all") in an autoused fixture on my side.

Done, and I moved these to the root `nilearn/conftest.py`. I figure a < 1 us overhead on each test run is acceptable. But if it's not, I can move the `plt.close('all')` autoused figure back into a new `nilearn/plotting/tests/conftest.py` or `nilearn/plotting/conftest.py` file

jeromedockes commented:
> Done, and I moved these to the root `nilearn/conftest.py`. I figure a < 1 us overhead on each test run is acceptable. But if it's not, I can move the `plt.close('all')` autoused figure back into a new `nilearn/plotting/tests/conftest.py` or `nilearn/plotting/conftest.py` file

Thanks! I think we can keep it in nilearn/conftest.py

KamalakerDadi commented on nilearn/conftest.py:
```diff
@@ -15,6 +15,22 @@
+def pytest_configure(config):
+    """Use Agg so that no figures pop up."""
+    if matplotlib is not None:
+        matplotlib.use('Agg', force=True)
```
I hope we will not be having issues as in https://github.com/nilearn/nilearn/issues/2560

larsoner replied:
The expectation for running tests should be different from the expectation when importing a library four actual use. In test running it's fine to set the backend. We've done it this way for years in MNE-Python without issue.

bthirion reviewed (approved):
Thx!

bthirion commented:
Merging, thx.
